src/OpenLoco/S5/SawyerStream.cpp

```diff
@@ -0,0 +1,488 @@
+}
+
+void SawyerStreamWriter::writeChecksum()
+{
+    _stream.write(reinterpret_cast<const char*>(&_checksum), sizeof(_checksum));
+}
+
+void SawyerStreamWriter::close()
+{
+    _stream.close();
+}
+
+stdx::span<uint8_t const> SawyerStreamWriter::encode(SawyerEncoding encoding, stdx::span<uint8_t const> data)
+{
+    switch (encoding)
+    {
+        case SawyerEncoding::uncompressed:
+            return data;
+        case SawyerEncoding::runLengthSingle:
+            _encodeBuffer.clear();
+            _encodeBuffer.reserve(data.size());
+            encodeRunLengthSingle(_encodeBuffer, data);
+            return _encodeBuffer.getSpan();
+        case SawyerEncoding::runLengthMulti:
+            _encodeBuffer.clear();
+            _encodeBuffer.reserve(data.size());
+            encodeRunLengthMulti(_encodeBuffer, data);
+
+            _encodeBuffer2.clear();
+            _encodeBuffer2.reserve(_encodeBuffer.size());
+            encodeRunLengthSingle(_encodeBuffer2, _encodeBuffer.getSpan());
+            return _encodeBuffer2.getSpan();
+        case SawyerEncoding::rotate:
+            _encodeBuffer.clear();
+            _encodeBuffer.reserve(data.size());
+            encodeRotate(_encodeBuffer, data);
+            return _encodeBuffer.getSpan();
+        default:
+            throw std::runtime_error(exceptionUnknownEncoding);
+    }
+}
+
+void SawyerStreamWriter::encodeRunLengthSingle(FastBuffer& buffer, stdx::span<uint8_t const> data)
+{
+    auto src = data.data();
+    auto srcEnd = src + data.size();
+    auto srcNormStart = src;
+    uint8_t count = 0;
+
+    while (src < srcEnd - 1)
+    {
+        if ((count != 0 && src[0] == src[1]) || count > 125)
+        {
+            buffer.push_back(count - 1);
+            buffer.push_back(srcNormStart, count);
+            srcNormStart += count;
+            count = 0;
+        }
+        if (src[0] == src[1])
+        {
+            for (; count < 125 && src + count < srcEnd; count++)
+            {
+                if (src[0] != src[count])
+                {
+                    break;
+                }
+            }
+            buffer.push_back(257 - count);
+            buffer.push_back(src[0]);
+            src += count;
+            srcNormStart = src;
+            count = 0;
+        }
+        else
+        {
+            count++;
+            src++;
+        }
+    }
+    if (src == srcEnd - 1)
+    {
+        count++;
+    }
+    if (count != 0)
+    {
+        buffer.push_back(count - 1);
+        buffer.push_back(srcNormStart, count);
+    }
+}
+
+void SawyerStreamWriter::encodeRunLengthMulti(FastBuffer& buffer, stdx::span<uint8_t const> data)
+{
+    auto src = data.data();
+    auto srcLen = data.size();
+    if (srcLen == 0)
+        return;
+
+    // Need to emit at least one byte, otherwise there is nothing to repeat
+    buffer.push_back(255);
+    buffer.push_back(src[0]);
+
+    // Iterate through remainder of the source buffer
+    for (size_t i = 1; i < srcLen;)
```

Comment:
Could we split this up into a number of smaller functions. One for finding the best repeat. One for applying the repeat on the data...